docs: Streaming guide (#90779)
vercel/next.js: docs/01-app/02-guides/streaming.mdx

---
title: Streaming
nav_title: Streaming
description: Learn how streaming works in Next.js and how to use it to progressively render UI as data becomes available.
related:
  description: Related API references and guides.
  links:
    - app/api-reference/file-conventions/loading
    - app/getting-started/fetching-data
    - app/getting-started/linking-and-navigating
    - app/guides/self-hosting
---

## What is streaming?

> In traditional server-side rendering, the server produces the full HTML document before sending anything. A single slow database query or API call holds up the entire page. Streaming changes this by using [chunked transfer encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Transfer-Encoding) to send parts of the response as they become ready. The browser starts rendering HTML while the server is still generating the rest.

Reviewer: Great intro.

```suggestion
In traditional server-side rendering, the server produces the full HTML document before sending anything. A single slow database query or API call can block the entire page. Streaming changes this by using [chunked transfer encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Transfer-Encoding) to send parts of the response as they become ready. The browser starts rendering HTML while the server is still generating the rest.
```

This is especially impactful for pages that combine fast static content (headers, navigation, layout) with slower dynamic content (personalized data, analytics, recommendations). The static parts can be prerendered and served from a CDN, painting instantly, while the dynamic parts stream in from the server as they become ready.

> React's server renderer produces HTML in chunks aligned to `<Suspense>` boundaries. Next.js wires this into the App Router so streaming works out of the box.

Reviewer: I think we could clarify here on what "React server renderer" means (or link to a resource) as it's not a term you often see React using externally. Maybe name a specific React API `renderToXStream` or we could mention how React supports streaming and progressive hydration, with Suspense is used as a "marker" or boundary for chunks, or something among those lines. 

```suggestion
React's server renderer produces HTML in chunks aligned with `<Suspense>` boundaries. Next.js integrates this into the App Router so streaming works without additional configuration. 
```

      <div className="h-4 w-2/3 bg-gray-200 rounded" />
    </div>
  )
}
```

Behind the scenes, `loading.js` is nested inside `layout.js` and wraps `page.js` in a `<Suspense>` boundary:

<Image
  alt="loading.js overview"
  srcLight="/docs/light/loading-overview.png"
  srcDark="/docs/dark/loading-overview.png"
  width="1600"
  height="768"
/>

This means:

- The layout renders immediately as part of the static shell.
- The loading skeleton is shown instantly as the Suspense fallback.
- When the page component finishes loading, its HTML replaces the skeleton.

`loading.js` is useful when there's nothing meaningful to show until the page's data resolves. If the page needs to await data before it can render anything, a full-page skeleton is a reasonable fallback.

See the [`loading.js` API reference](/docs/app/api-reference/file-conventions/loading) for more details.

## Granular streaming with `<Suspense>`

`<Suspense>` lets you control exactly which parts of the page stream independently. Instead of a full-page skeleton, you can push fallbacks down into specific sections so the static shell includes more real content.

### Parallel streaming with sibling boundaries

> Wrap each independent section in its own `<Suspense>` boundary. They resolve and stream in whatever order they finish. The fast ones don't wait for the slow ones:

Reviewer: ```suggestion
Wrap each independent section in its own `<Suspense>` boundary. They resolve and stream in whatever order they finish, without blocking each other.
```

import { Suspense } from 'react'
import { StatsChart } from './stats-chart'

async function getStats() {
  const res = await fetch('https://api.example.com/stats')
  return res.json()
}

export default function Dashboard() {
  const statsPromise = getStats()

  return (
    <Suspense fallback={<p>Loading chart...</p>}>
      <StatsChart dataPromise={statsPromise} />
    </Suspense>
  )
}
```

```tsx filename="app/dashboard/stats-chart.tsx" switcher
'use client'

import { use } from 'react'

type Stats = { revenue: number; orders: number }

export function StatsChart({ dataPromise }: { dataPromise: Promise<Stats> }) {
  const stats = use(dataPromise)

  return <div>{/* render chart with stats */}</div>
}
```

```jsx filename="app/dashboard/stats-chart.js" switcher
'use client'

import { use } from 'react'

export function StatsChart({ dataPromise }) {
  const stats = use(dataPromise)

  return <div>{/* render chart with stats */}</div>
}
```

The Server Component starts the fetch and passes the unresolved promise to the Client Component. The HTML for the `<Suspense>` fallback is sent immediately. When the promise resolves, React streams the completed component to the client.

> ### Sharing a promise across the client tree

Reviewer: Should we add an example for this?

Author: I think we can tackle this further here: https://linear.app/vercel/issue/DOC-6143/streaming-data-to-the-client
 
I do think I jumped too far ahead here though, should start with, you kick off a promise server side, and React can transfer it across the boundary into the client. Then, you can think of making it available through Context providers and such.

    },
  })
}
```

Visit this route directly in the browser or with `curl` to see chunks arrive one at a time:

```bash
curl http://localhost:3000/api/stream
```

You can also stream files without loading them entirely into memory. Use `FileHandle.readableWebStream()` to get a Web `ReadableStream` directly from a file:

```ts filename="app/api/download/route.ts" switcher
import { open } from 'node:fs/promises'

export async function GET() {
  const file = await open('/path/to/large-file.csv')

  return new Response(file.readableWebStream(), {
    headers: {
      'Content-Type': 'text/csv',
      'Content-Disposition': 'attachment; filename="data.csv"',
    },
  })
}
```

```js filename="app/api/download/route.js" switcher
import { open } from 'node:fs/promises'

export async function GET() {
  const file = await open('/path/to/large-file.csv')

  return new Response(file.readableWebStream(), {
    headers: {
      'Content-Type': 'text/csv',
      'Content-Disposition': 'attachment; filename="data.csv"',
    },
  })
}
```

See the [Route Handler API reference](/docs/app/api-reference/file-conventions/route) for more details on building streaming endpoints.

> ## Streaming and Web Vitals

Reviewer: To help place the user, you could add a sentence here on web vitals and link to the Google docs.

  return (
    <Suspense fallback={<p>Loading post...</p>}>
      <PostContent slug={slug} />
    </Suspense>
  )
}
```

> **Good to know:** You can also reject requests early using [`proxy`](/docs/app/api-reference/file-conventions/proxy) (for redirects, rewrites, or returning a response) or [`next.config.js` redirects](/docs/app/api-reference/config/next-config-js/redirects). Both run before the page renders, so HTTP status codes are still available.

### Metadata and bots

[`generateMetadata`](/docs/app/api-reference/functions/generate-metadata) resolves before streaming begins for bots that only scrape static HTML (such as Twitterbot or Slackbot). For full browsers and capable crawlers, metadata can [stream](/docs/app/api-reference/functions/generate-metadata#streaming-metadata) alongside the page content.

Next.js automatically detects user agents to choose the right behavior. You can customize which bots receive blocking metadata with the [`htmlLimitedBots`](/docs/app/api-reference/config/next-config-js/htmlLimitedBots) configuration option.

See the [`loading.js` SEO section](/docs/app/api-reference/file-conventions/loading#seo) for more details.

> ## Streaming through the infrastructure stack

Reviewer: Maybe rephrase this was things that can affect / impact streaming

```

The `<template id="B:0">` markers are the Suspense fallback placeholders. When a boundary resolves, React streams a `<div hidden id="S:0">` containing the completed HTML and a script that swaps it into the page. The timestamps show each boundary resolving independently.

> **Good to know:** The `Accept-Encoding: identity` header disables compression so chunks are not buffered by the compression layer.

### Platform support

| Deployment Option                                                   | Supported         |
| ------------------------------------------------------------------- | ----------------- |
| [Node.js server](/docs/app/getting-started/deploying#nodejs-server) | Yes               |
| [Docker container](/docs/app/getting-started/deploying#docker)      | Yes               |
| [Static export](/docs/app/getting-started/deploying#static-export)  | No                |
| [Adapters](/docs/app/getting-started/deploying#adapters)            | Platform-specific |

See the [Self-Hosting guide](/docs/app/guides/self-hosting#streaming-and-suspense) for detailed configuration instructions.

## How the concepts connect

> The trigger is **your code**. The framework requires explicitness about three kinds of work:

Reviewer: This section is similar to the "How App Router streams". Perhaps we can combine them?

Author: I initially had, and want this to be, the tl;dr 

I think I can cut/move some bits to their respective section, and still keep a nice tl;dr

3. **Streaming** delivers the static shell immediately over HTTP. As each boundary resolves, React streams the completed HTML along with a small script that swaps it into the page in-place, without waiting for the full JavaScript bundle.
4. **Caching directives** like [`"use cache"`](/docs/app/api-reference/directives/use-cache) let you push work back into the shell. With [Cache Components](/docs/app/getting-started/caching), cached components are prerendered at build time and served from the edge. Their output becomes part of the static shell. Uncached components stream at request time. This means you control the boundary between instant and streamed content by choosing which components use `"use cache"`.

The key decisions are **what to cache** and **where to place Suspense boundaries**. Cache what you can to grow the static shell. Push dynamic access down to the components that need it, and wrap those components in `<Suspense>`. Everything else becomes part of the shell.

> ## Example

Reviewer: Move the example to the beginning so people can see it in action as they read through the sections.
